counterpartylib/lib/transaction.py

```diff
@@ -487,4 +491,199 @@ def construct (db, tx_info, encoding='auto',
+
+    raise exceptions.UnpackError("Invalid data")
+
+def get_outputs_info(tx_hex): 
+    # Decode transaction binary.
+    ctx = bitcoinlib.core.CTransaction.deserialize(binascii.unhexlify(tx_hex))
+
+    # Ignore coinbase transactions.
+    if ctx.is_coinbase():
+        raise exceptions.TransactionError('coinbase transaction')
+
+    # Get destinations and data outputs.
+    destinations, btc_amount, data, change = [], 0, b'', (None, 0)
+
+    outputs_info = {
+        'data': b'',
+        'data_value': 0,
+        'data_output_count': 0,
+        'data_output_type': None,
+        'total_value': 0,
+        'destinations': {},
+        'change': (None, 0)
+    }
+    for vout in ctx.vout:
+        # change always last
+        if outputs_info['change'][0] is not None:
+            raise exceptions.TransactionError('constructed transaction contains more than one change output')
+
+        new_destination, new_data, output_type = blocks.decode_output(vout, ctx)
+
+        # All destinations come before all data.
+        if not outputs_info['data'] and not new_data:
+            outputs_info['destinations'][new_destination] = vout.nValue
+        else:
+            if new_destination:     # Change.
+                outputs_info['change'] = (new_destination, vout.nValue)
+            else:                   # Data.
+                outputs_info['data'] += new_data
+                outputs_info['data_value'] += vout.nValue
+                outputs_info['data_output_count'] += 1
+                if outputs_info['data_output_type'] is not None and outputs_info['data_output_type'] != output_type:
+                    raise exceptions.TransactionError('constructed transaction contains invalid data outputs')
+                outputs_info['data_output_type'] = output_type
+        outputs_info['total_value'] += vout.nValue
+
+    return outputs_info
+
+def check_outputs(method, params, tx_hex):
+    tx_info = get_outputs_info(tx_hex)
+    #print(tx_info)
+    asked_message_type = method.split('_')[1]
+
+    if tx_info['data']:
+        
+        data_hex = binascii.hexlify(tx_info['data'])
+        data = raw_unpack_data(data_hex)
+        #print(params)
+        #print(data)
+
+        # check message type
+        returned_message_type = data.pop('message')
+        if asked_message_type != returned_message_type:
+            raise exceptions.TransactionError('invalid message type')
+
+        # asset_id to asset name (hackish)
+        asset_param_list = {
+            'asset_id': 'asset', 
+            'give_id': 'give_asset', 
+            'get_id': 'get_asset',
+            'dividend_asset_id': 'dividend_asset'
+        }
+        for asset_param in asset_param_list:
+            if asset_param in data:
+                asset_id = data.pop(asset_param)
+                param_name = asset_param_list[asset_param]
+                try:
+                    data[param_name] = util.generate_asset_name(asset_id, block_index=333501)
```

Comment:
hum.. I don't think, for instance we can have a `send` transaction for an "old" asset.